Add-on library for recoverable signatures

evgeny-osipenko opened the pull request:
This is a continuation of PR #44, in accordance with the multi-package approach outlined in issue #43.

This has the same functionality as PR #42, but organized differently. Whereas #42 simply adds the recovery functions to the single module; this one has them as a separate package.

jprupp reviewed (changes requested):
Please rebase against latest master, remove `README.md` file from pacakge, and include a copy of the `LICENSE` file instead of a symlink, as they break some Haskell tools.

evgeny-osipenko commented:
Removed `README.md` from the filesystem and from `package.yaml`, changed `LICENSE` from a symlink to a full copy (should it still say the year "2020" btw?), and also removed a couple of redundant imports from the code itself after checking it with `-Wall`.

evgeny-osipenko commented:
@jprupp I made the changes you asked for, is there anything else left to do before this PR can be merged in?

jprupp commented on secp256k1-haskell-recovery/package.yaml:
```diff
@@ -0,0 +1,44 @@
+synopsis: Bindings for recoverable signatures feature of secp256k1
+description: Sign and verify recoverable signatures using the secp256k1 library.
+category: Crypto
+author: Jean-Pierre Rupp
```
This is your package. You can put your name here instead of mine.

evgeny-osipenko replied:
Done

jprupp commented on secp256k1-haskell-recovery/package.yaml:
```diff
@@ -0,0 +1,44 @@
+category: Crypto
+author: Jean-Pierre Rupp
+maintainer: jprupp@protonmail.ch
+copyright: (c) 2017 Jean-Pierre Rupp
```
Ditto.

evgeny-osipenko replied:
Done

jprupp reviewed (changes requested):
It looks good. Just set yourself as the author and I can merge it.

evgeny-osipenko commented:
Changed the names to myself and amended the commit.

evgeny-osipenko commented:
Actually pushed the changes now.

jprupp reviewed (changes requested):
Test Fails:

```
secp256k1-haskell-recovery> test (suite: spec)


Crypto.Secp256k1.Recovery
  recovery
    recovers public keys [✔]     
      +++ OK, passed 100 tests.
    recovers key from signed message [✔]     
      +++ OK, passed 100 tests.
    does not recover bad public keys [✔]     
      +++ OK, passed 100 tests.
    detects bad recoverable signature [✔]     
      +++ OK, passed 100 tests.
    serializes compact recoverable signature [✔]     
      +++ OK, passed 100 tests.
    shows and reads recoverable signature [✔]     
      +++ OK, passed 100 tests.
    reads recoverable signature from string [✔]     
      +++ OK, passed 100 tests.
    produces the expected signature [✘]    
    recovers the expected pub key [✔]
      +++ OK, passed 1 test.

Failures:

  test/Crypto/Secp256k1/RecoverySpec.hs:107:3: 
  1) Crypto.Secp256k1.Recovery.recovery produces the expected signature
       Falsifiable (after 1 test):
       
       produced signature matches
       expected: "02559ab98a8908ba4cf0f914eb8b66651405ab69ab7c461dd140e40baa1b5e1d2ef6ac88a13e2226f76a9d8d49bb9cf3061dac1364c0dfe7b69cd165a8d08f0700"
        but got: "d0205e402f5c0bc8bbbcf08b3f0912b055fc3998ef0d090c2dec52f98695c4e01cd55228c2370ffe52826b3f2bd82e42fcf94e8943cd402f39644935b2c7c08b00"

  To rerun use: --match "/Crypto.Secp256k1.Recovery/recovery/produces the expected signature/"

Randomized with seed 1890685244

Finished in 6.2283 seconds
9 examples, 1 failure
```
